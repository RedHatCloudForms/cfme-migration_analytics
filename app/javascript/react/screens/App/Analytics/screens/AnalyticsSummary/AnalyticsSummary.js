import React from 'react';
import PropTypes from 'prop-types';
import { Spinner } from 'patternfly-react';
import { VM_SUMMARY_REPORT_FILTERS, FINISHED, OK } from '../../constants';

// TODO sort out data we need from the mocks and write helpers to calculate it
// TODO build accordion showing summary data
// TODO provider selection?
// TODO next page?

class AnalyticsSummary extends React.Component {
  constructor(props) {
    super(props);
    this.taskFetchTimeout = null;
  }

  componentDidMount() {
    const { fetchReportsAction } = this.props;
    fetchReportsAction(VM_SUMMARY_REPORT_FILTERS); // Kicks off the chain of action calls in componentDidUpdate
  }

  componentDidUpdate(prevProps) {
    const {
      vmSummaryReport,
      runReportAction,
      vmSummaryReportRun,
      fetchTaskAction,
      isFetchingVmSummaryReportTask,
      vmSummaryReportTask,
      fetchResultAction,
      vmSummaryReportResult,
      calculateSummaryDataAction
    } = this.props;

    // Once we've found the report href, run it.
    if (!prevProps.vmSummaryReport && vmSummaryReport) {
      runReportAction(vmSummaryReport.href);
      return;
    }

    // Once we have a task href for the running report, fetch it.
    if (!prevProps.vmSummaryReportRun && vmSummaryReportRun) {
      fetchTaskAction(vmSummaryReportRun.task_href);
      return;
    }

    // If we fetched an unfinished task, wait and fetch it again.
    if (
      prevProps.isFetchingVmSummaryReportTask &&
      !isFetchingVmSummaryReportTask &&
      vmSummaryReportTask &&
      vmSummaryReportTask.state !== FINISHED
    ) {
      this.taskFetchTimeout = setTimeout(() => fetchTaskAction(vmSummaryReportRun.task_href), 1000);
      return;
    }

    // Once we have a successfully finished task, fetch its result.
    if (
      (!prevProps.vmSummaryReportTask || prevProps.vmSummaryReportTask.state !== FINISHED) &&
      vmSummaryReportTask &&
      vmSummaryReportTask.state === FINISHED &&
      vmSummaryReportTask.status === OK
    ) {
      clearTimeout(this.fetchTaskTimeout); // Just in case.
      fetchResultAction(vmSummaryReportRun.result_href);
      return;
    }

    // Once we have a report result, process it for rendering.
    if (!prevProps.vmSummaryReportResult && vmSummaryReportResult) {
      calculateSummaryDataAction(vmSummaryReportResult);
    }
  }

  render() {
    const { summaryData } = this.props;
    if (!summaryData) {
      return (
        <div className="large-spinner">
          <Spinner loading size="lg" inline />
          <h3>{__('Examining virtualization providers')}</h3>
        </div>
      );
    }

    // TODO handle case where vmSummaryReportTask has an error

    return (
      <React.Fragment>
        <h1>TODO</h1>
        <pre>{JSON.stringify(summaryData, 2)}</pre>
      </React.Fragment>
    );
  }
}

AnalyticsSummary.propTypes = {
  fetchReportsAction: PropTypes.func,
  vmSummaryReport: PropTypes.shape({
    href: PropTypes.string,
    name: PropTypes.string,
    rpt_group: PropTypes.string
  }),
  runReportAction: PropTypes.func,
  vmSummaryReportRun: PropTypes.shape({
    href: PropTypes.string,
    result_href: PropTypes.string,
    task_href: PropTypes.string
  }),
  fetchTaskAction: PropTypes.func,
  isFetchingVmSummaryReportTask: PropTypes.bool,
  vmSummaryReportTask: PropTypes.shape({
    id: PropTypes.string,
    state: PropTypes.string,
    status: PropTypes.string
  }),
  fetchResultAction: PropTypes.func,
  vmSummaryReportResult: PropTypes.shape({
    result_set: PropTypes.arrayOf(
      PropTypes.shape({
        allocated_disk_storage: PropTypes.number,
        mem_cpu: PropTypes.number,
        cpu_total_cores: PropTypes.number,
        'ext_management_system.name': PropTypes.string,
        'ext_management_system.id': PropTypes.number
      })
    )
  }),
  calculateSummaryDataAction: PropTypes.func,
  summaryData: PropTypes.shape({
    total: PropTypes.shape({
      numProviders: PropTypes.number,
      numHypervisors: PropTypes.number,
      numVms: PropTypes.number,
      allocatedDiskSpace: PropTypes.number,
      allocatedMemory: PropTypes.number,
      numCpuCores: PropTypes.number
    }),
    providers: PropTypes.arrayOf(
      PropTypes.shape({
        numHypervisors: PropTypes.number,
        numVms: PropTypes.number,
        allocatedDiskSpace: PropTypes.number,
        allocatedMemory: PropTypes.number,
        numCpuCores: PropTypes.number
      })
    )
  })
};

export default AnalyticsSummary;
