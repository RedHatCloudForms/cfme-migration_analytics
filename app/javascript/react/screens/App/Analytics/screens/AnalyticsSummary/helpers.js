export const someProvidersExist = ({ providers }) => providers && providers.length > 0;

export const noRefreshErrorsExist = ({ providersWithRefreshErrors }) =>
  providersWithRefreshErrors && providersWithRefreshErrors.length === 0;

export const someProvidersAwaitingRefresh = ({ providersAwaitingRefresh }) =>
  providersAwaitingRefresh && providersAwaitingRefresh.length > 0;

export const noProvidersAwaitingRefresh = ({ providersAwaitingRefresh }) =>
  providersAwaitingRefresh && providersAwaitingRefresh.length === 0;

export const providersRefreshed = props =>
  someProvidersExist(props) && noRefreshErrorsExist(props) && noProvidersAwaitingRefresh(props);
