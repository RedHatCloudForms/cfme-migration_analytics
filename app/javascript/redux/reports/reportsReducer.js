import Immutable from 'seamless-immutable';
import { FETCH_REPORTS, RUN_REPORT, FETCH_TASK, FETCH_RESULT } from './constants';

export const initialState = Immutable({
  numReportFetchesPending: 0,
  errorFetchingReports: null,
  reports: [],
  numReportRunsPending: 0,
  errorRunningReport: null,
  reportRuns: [], // TODO index this by id/href instead?
  fetchingTaskHrefs: [],
  errorFetchingTask: null,
  tasksById: {},
  fetchingResultHrefs: [],
  errorFetchingResult: null,
  resultsById: {}
});

export default (state = initialState, action) => {
  switch (action.type) {
    // TODO reset to initial state when the main analytics page mounts? otherwise will we get `reports` building up?
    case `${FETCH_REPORTS}_PENDING`:
      return state.set('numReportFetchesPending', state.numReportFetchesPending + 1).set('errorFetchingReports', null);
    case `${FETCH_REPORTS}_FULFILLED`:
      return state
        .set('numReportFetchesPending', state.numReportFetchesPending - 1)
        .set('errorFetchingReports', null)
        .set('reports', [...state.reports, ...action.payload.data.resources]);
    case `${FETCH_REPORTS}_REJECTED`:
      return state
        .set('numReportFetchesPending', state.numReportFetchesPending - 1)
        .set('errorFetchingReports', action.payload);

    case `${RUN_REPORT}_PENDING`:
      return state.set('numReportRunsPending', state.numReportRunsPending + 1).set('errorRunningReport', null);
    case `${RUN_REPORT}_FULFILLED`:
      return state
        .set('numReportRunsPending', state.numReportRunsPending - 1)
        .set('errorRunningReport', null)
        .set('reportRuns', [
          ...state.reportRuns.filter(run => run.href !== action.payload.data.href),
          action.payload.data
        ]);
    case `${RUN_REPORT}_REJECTED`:
      return state
        .set('numReportRunsPending', state.numReportRunsPending - 1)
        .set('errorRunningReport', action.payload);

    case `${FETCH_TASK}_PENDING`:
      return state
        .set('fetchingTaskHrefs', [...state.fetchingTaskHrefs, action.meta.href])
        .set('errorFetchingTask', null);
    case `${FETCH_TASK}_FULFILLED`:
      return state
        .set('fetchingTaskHrefs', state.fetchingTaskHrefs.filter(href => href !== action.meta.href))
        .set('errorFetchingTask', null)
        .set('tasksById', { ...state.tasksById, [action.payload.data.id]: action.payload.data });
    case `${FETCH_TASK}_REJECTED`:
      return state
        .set('fetchingTaskHrefs', state.fetchingTaskHrefs.filter(href => href !== action.meta.href))
        .set('errorFetchingTask', action.payload);

    case `${FETCH_RESULT}_PENDING`:
      return state
        .set('fetchingResultHrefs', [...state.fetchingResultHrefs, action.meta.href])
        .set('errorFetchingResult', null);
    case `${FETCH_RESULT}_FULFILLED`:
      return state
        .set('fetchingResultHrefs', state.fetchingResultHrefs.filter(href => href !== action.meta.href))
        .set('errorFetchingResult', null)
        .set('resultsById', { ...state.resultsById, [action.payload.data.id]: action.payload.data });
    case `${FETCH_RESULT}_REJECTED`:
      return state
        .set('fetchingResultHrefs', state.fetchingResultHrefs.filter(href => href !== action.meta.href))
        .set('errorFetchingResult', action.payload);

    default:
      return state;
  }
};
