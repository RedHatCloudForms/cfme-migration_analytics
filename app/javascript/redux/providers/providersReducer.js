import Immutable from 'seamless-immutable';
import { functionLookupReducer, getHandlersForFetchResourcesActions } from '../helpers';
import { FETCH_PROVIDERS } from './constants';

const fetchProviders = getHandlersForFetchResourcesActions(
  FETCH_PROVIDERS,
  'isFetchingProviders',
  'errorFetchingProviders',
  'providers'
);

export const initialState = Immutable({
  ...fetchProviders.initialState
});

const actionHandlers = {
  ...fetchProviders.actionHandlers
};

export default functionLookupReducer(initialState, actionHandlers);
