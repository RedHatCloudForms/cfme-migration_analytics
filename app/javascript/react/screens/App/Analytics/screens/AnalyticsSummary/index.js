import { connect } from 'react-redux';
import AnalyticsSummary from './AnalyticsSummary';
import { fetchReportsAction, runReportAction } from '../../../../../../redux/reports/reportsActions';
import { selectReportByFilterValues, selectRunByReport } from '../../../../../../redux/reports/reportsSelectors';
import { VM_SUMMARY_REPORT_FILTERS } from '../../constants';

const mapStateToProps = state => {
  const vmSummaryReport = selectReportByFilterValues(state, VM_SUMMARY_REPORT_FILTERS);
  return {
    vmSummaryReport,
    lastVmSummaryReportRun: selectRunByReport(state, vmSummaryReport)
  };
};

export default connect(
  mapStateToProps,
  { fetchReportsAction, runReportAction }
)(AnalyticsSummary);
