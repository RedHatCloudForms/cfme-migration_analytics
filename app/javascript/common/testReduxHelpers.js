import React from 'react';
import thunkMiddleware from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import configureMockStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

const middlewares = [thunkMiddleware, promiseMiddleware];

export const generateStore = (reducers, initialState) =>
  createStore(combineReducers(reducers), initialState, applyMiddleware(...middlewares));

export const mockStore = initialState => configureMockStore(middlewares)(initialState);

export const mountWithTestStore = (reducers, initialState, children) => {
  const store = generateStore({ migrationAnalytics: combineReducers(reducers) }, { migrationAnalytics: initialState });
  return mount(<Provider store={store}>{children}</Provider>);
};

export const mockDispatch = actionCreator => {
  const dispatch = jest.fn();
  actionCreator(dispatch);
  return dispatch.mock.calls[0][0];
};
