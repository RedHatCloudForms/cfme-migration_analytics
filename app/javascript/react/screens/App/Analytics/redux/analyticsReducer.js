import Immutable from 'seamless-immutable';
import {
  CALCULATE_SUMMARY_DATA,
  SELECT_PROVIDERS,
  SELECT_DETAILED_DATA,
  FETCH_MANIFEST_INFO,
  TOGGLE_MANIFEST_UPDATE_MODAL,
  START_INVENTORY_BUNDLE,
  FETCH_BUNDLE_TASK,
  RESET_DATA_COLLECTION_STATE,
  CHANGE_MANIFEST
} from './constants';
import { functionLookupReducer, getHandlersForBasicFetchActions } from '../../../../../redux/helpers';
import { calculateSummaryData } from './helpers';
import { RESET_ALL_STATE } from '../../../../../redux/common/constants';

const fetchManifestInfo = getHandlersForBasicFetchActions(
  FETCH_MANIFEST_INFO,
  'isFetchingManifestInfo',
  'errorFetchingManifestInfo',
  'manifestInfo'
);

const changeManifest = getHandlersForBasicFetchActions(
  CHANGE_MANIFEST,
  'isChangingManifest',
  'errorChangingManifest',
  'changeManifestResult'
);

const startInventoryBundle = getHandlersForBasicFetchActions(
  START_INVENTORY_BUNDLE,
  'isStartingBundle',
  'errorStartingBundle',
  'bundleStartResult'
);

const fetchBundleTask = getHandlersForBasicFetchActions(
  FETCH_BUNDLE_TASK,
  'isFetchingBundleTask',
  'errorFetchingBundleTask',
  'bundleTask'
);

export const initialState = Immutable({
  ...fetchManifestInfo.initialState,
  ...changeManifest.initialState,
  manifestUpdateModalVisible: false,
  summaryData: null,
  selectedProviders: [],
  detailedDataSelected: false,
  ...startInventoryBundle.initialState,
  ...fetchBundleTask.initialState
});

const actionHandlers = {
  ...fetchManifestInfo.actionHandlers,
  ...changeManifest.actionHandlers,
  [TOGGLE_MANIFEST_UPDATE_MODAL]: (state, action) =>
    state.set('manifestUpdateModalVisible', action.show !== null ? action.show : !state.manifestUpdateModalVisible),
  [CALCULATE_SUMMARY_DATA]: (state, action) => state.set('summaryData', calculateSummaryData(action.results)),
  [SELECT_PROVIDERS]: (state, action) => state.set('selectedProviders', action.selectedProviders),
  [SELECT_DETAILED_DATA]: (state, action) => state.set('detailedDataSelected', action.detailedDataSelected),
  ...startInventoryBundle.actionHandlers,
  ...fetchBundleTask.actionHandlers,
  [RESET_DATA_COLLECTION_STATE]: state =>
    Immutable.merge(state, {
      ...startInventoryBundle.initialState,
      ...fetchBundleTask.initialState
    }),
  [RESET_ALL_STATE]: () => initialState
};

export default functionLookupReducer(initialState, actionHandlers);
