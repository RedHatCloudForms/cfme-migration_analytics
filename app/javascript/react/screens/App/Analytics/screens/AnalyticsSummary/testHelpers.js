import { expectOnlySpecificActions } from '../../../../../../common/testReduxHelpers';

export const expectInitialFetchActions = (calledTimes, props) => {
  expect(props.fetchProvidersAction).toHaveBeenCalledTimes(calledTimes * 1);
  expect(props.fetchReportsAction).toHaveBeenCalledTimes(calledTimes * 2);
};

export const expectNoActionsOtherThan = (expectedActions, props) => {
  const allActions = [
    'fetchProvidersAction',
    'fetchReportsAction',
    'runReportAction',
    'fetchTaskAction',
    'fetchResultAction',
    'calculateSummaryDataAction',
    'resetAllStateAction'
  ];
  expectOnlySpecificActions({ allActions, expectedActions, props });
};

export const expectNoActions = props => expectNoActionsOtherThan([], props);

export const expectOnlyInitialFetchActions = (props, initialFetchTimes = 1) => {
  expectInitialFetchActions(initialFetchTimes, props);
  expectNoActionsOtherThan(['fetchProvidersAction', 'fetchReportsAction'], props);
};
