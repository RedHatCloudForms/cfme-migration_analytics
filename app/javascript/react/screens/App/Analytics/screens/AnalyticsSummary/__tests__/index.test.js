import React from 'react';

import { mountWithTestStore } from '../../../../../../../common/testReduxHelpers';
import analyticsReducer from '../../../redux/analyticsReducer';
import providersReducer from '../../../../../../../redux/providers/providersReducer';
import reportsReducer from '../../../../../../../redux/reports/reportsReducer';

import AnalyticsSummary from '../AnalyticsSummary';
import AnalyticsSummaryContainer from '../index';

test('some component should mount with mapStateToProps reduced', () => {
  const wrapper = mountWithTestStore(
    { providers: providersReducer, reports: reportsReducer, analytics: analyticsReducer },
    {},
    <AnalyticsSummaryContainer />
  );
  const component = wrapper.find(AnalyticsSummary);
  expect(component.props()).toMatchSnapshot();
});
