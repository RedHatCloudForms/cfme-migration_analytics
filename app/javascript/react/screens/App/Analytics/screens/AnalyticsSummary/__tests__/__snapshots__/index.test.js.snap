// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`some component should mount with mapStateToProps reduced 1`] = `
Object {
  "calculateSummaryDataAction": [Function],
  "envSummaryReport": null,
  "envSummaryReportResult": null,
  "envSummaryReportRun": null,
  "envSummaryReportTask": null,
  "errorFetchingProviders": null,
  "errorFetchingReports": null,
  "fetchProvidersAction": [Function],
  "fetchReportsAction": [Function],
  "fetchResultAction": [Function],
  "fetchTaskAction": [Function],
  "isFetchingEnvSummaryReportTask": false,
  "isFetchingProviders": true,
  "isFetchingVmSummaryReportTask": false,
  "providers": null,
  "providersAwaitingRefresh": null,
  "providersWithRefreshErrors": null,
  "resetAllStateAction": [Function],
  "runReportAction": [Function],
  "summaryData": null,
  "vmSummaryReport": null,
  "vmSummaryReportResult": null,
  "vmSummaryReportRun": null,
  "vmSummaryReportTask": null,
}
`;
